src/history/hs_rec.c

```diff
@@ -680,16 +646,25 @@ __wt_hs_insert_updates(
 }
 
 /*
- * __wt_hs_delete_key_from_ts --
- *     Delete history store content of a given key from a timestamp.
+ * __wt_hs_delete_reinsert_key_from_ts --
+ *     Delete history store content of a given key from a timestamp and optionally reinsert them
+ *     with ts-1 timestamp.
  */
 int
-__wt_hs_delete_key_from_ts(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, uint32_t btree_id,
-  const WT_ITEM *key, wt_timestamp_t ts, bool reinsert)
+__wt_hs_delete_reinsert_key_from_ts(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor,
+  uint32_t btree_id, const WT_ITEM *key, wt_timestamp_t ts, bool reinsert)
 {
     WT_DECL_RET;
+    WT_ITEM hs_key;
+    wt_timestamp_t hs_ts;
+    uint64_t hs_counter;
+    uint32_t hs_btree_id;
     bool hs_read_committed;
 
+    /* If we will delete all the updates of the key from the history store, we should not reinsert
+     * any update. */
+    WT_ASSERT(session, ts > WT_TS_NONE || !reinsert);
+
     hs_read_committed = F_ISSET(hs_cursor, WT_CURSTD_HS_READ_COMMITTED);
     if (!hs_read_committed)
         F_SET(hs_cursor, WT_CURSTD_HS_READ_COMMITTED);
```

Comment:
Do we also need to use the flag `WT_CURSTD_HS_READ_ALL` for this search_near scenario?

Reply:
Yes, why you think it's not needed? This is from your change. I'm just merging it.

Reply:
I didn't include the flag then. Can you review the need for flag here given we are refactoring out-of-order update code?

Reply:
What do you mean you didn't include? If you look at the code. The code is in the base commit.

Reply:
What I am saying is that I never analysed if we need to put flag `WT_CURSTD_HS_READ_ALL` here (before search near).

Reply:
I believe we should here.

Reply:
@quchenhao have you tried setting `WT_CURSTD_HS_READ_ALL` here?

Reply:
No. I was mistakenly think you are talking about `WT_CURSTD_HS_READ_COMMITTED` flag.

Reply:
@hbokhari I checked the old code and I think we need to set the flag here. However, I don't want to do this in this ticket. Can you create a separate ticket for this task?